accepted/00000-use-dnf-for-rpm-repos.md

```diff
@@ -0,0 +1,175 @@
+to handling modules in the future.
+
+As of today, DNF is available across RHEL/CentOS 7 and newer, Fedora, and
+(open)SUSE Linux 15.0 and newer (though PackageHub for SLE 15). Starting with
+openSUSE Leap 15.1, DNF supports RH/Fedora modules. So migrating to DNF will
+restore the capability to run Uyuni on CentOS and Fedora servers, while
+preserving the ability to run on openSUSE and SUSE Linux Enterprise 15.
+
+# Detailed design
+[design]: #detailed-design
+
+The below design information is derived from detailed design information from the "YUM replacement" RFC:
+
+### The `spacewalk-repo-sync` architecture
+
+![spacewalk-repo-sync diagram](images/reposync-diagram.png)
+
+### The purpose of a "ContentSource" plugin:
+
+Since there are multiple types of content sources (i.a. repositories types),
+`spacewalk-repo-sync` delegates the handling of each type to plugins. These "ContentSource" plugins must:
+
+- Get repository information and metadata from a repository (ContentSource).
+- Provide a list of packages, products and errata coming from that ContentSource. (ContentPackage)
+- Provide contact & download parameters to the Downloader Thread for downloading the packages.
+- Resolve package dependencies. Since `spacewalk-repo-sync` allows filtering (include/exclude) packages when syncing, package dependencies need to be calculated when running on this mode.
+
+With this information collected by the plugin, the `DownloaderThread` from `reposync`
+is able to perform the actual download of the necessary package files when `spacewalk-repo-sync` is called.
+
+Currently, `spacewalk-repo-sync` comes with the following plugins:
+- `yum_src`: RPM based repositories
+- `deb_src`: DEB based repositories
+- `uln_src`: Oracle ULN repositories (a wrapper of `yum_src` with custom authentication)
+
+### The "ContentSource" plugin API
+
+Mandatory methods:
+- `get_md_checksum_type`: Return the checksum_type of primary.xml
+- `get_products`: Return products metadata if any
+- `get_susedata`: Return suse metadata if any
+- `get_updates`: Return "updateinfo" / "patches" info
+- `get_groups`: Return repo groups if any
+- `get_file`:
+- `get_modules`: Return module metadata if any
+- `raw_list_packages`: Return the raw list of packages available after filtering.
+- `list_packages`: Return a list of `ContentPackage` rhn class for the available packages after filtering.
+- `clear_cache`: Clear cache files
+- `set_download_parameters`: Prepare the downloader params dictionary
+
+Yum specific methods we also need to cover:
+- `repomd_up_to_date`: Check if repomd.xml checksum is up-to-date.
+- `get_metadata_paths`: Simply load primary and updateinfo path from repomd.
+
+#### The "ContentPackage" class
+A "ContentSource" plugin needs to provide a list of "ContentPackage" objects to the reposync. This RHN class represents the package like by:
+
+- NEVRA
+- unique ID
+- checksum
+- checksum type
+
+### Replacing the custom "Zypper" + "libsolv" plugin with a "DNF" plugin.
+
```

Comment:
What about implementing a dnf module in parallel to yum.
Is there a reason why we would need to replace the existing one?

Reply:
I would imagine that for simplicity, it's easier to have one component that manages the rpm source stuff. It also makes the behaviors of the server consistent across whatever it is hosted on.

Maybe it makes sense to have an extra module for fetching legacy metadata (SLE 10 and older?), but it might be easy enough to handle everything through DNF.

Reply:
Having one code path that fails would indeed be easier. 😉  I know that there was a special case for <SLE10, but I don't know if this would/could be handled by DNF. @mcalmer, I think you mentioned that somewhere. Do you still remember what that was?

Reply:
@brejoc librepo has a stub for dealing with legacy susetags data (< SLE 10) which could be filled out, but maybe some custom urlgrabber code is sufficient here?

Reply:
I have the same feeling than @mcalmer here. Why not to implement this `dnf` plugin as an alternative plugin for managing RPM repositories?

Even if `dnf` is already packaged in PackageHub, bringing it as the default plugin for RPM means we would need also to support the DNF stack in our enterprise products.

Therefore, implementing this `dnf_src` plugin not as a replacement but as an alternative would be more interesting for us.

I'll elaborate this more in the main review comment.